### PR TITLE
Make CI checks to run on all pull requests

### LINKED ISSUE #3272: feat: make the CI work in with forks
### Describe the feature you'd like to request

Right now the CI doesn't work with fork, leading us to add a lot of people with write access. This is annoying and should be unnecessary

### Describe the solution you'd like to see

Make our PRs work with third party 

### Desribe alternate solutions

.

### Additional information

@Andarist looked at this and thinks it's because our push trigger in our yaml workflows.

Example: https://github.com/trpc/trpc/blob/22f268d1a16daa9154db2a456323366471fe30d7/.github/workflows/main.yml#L3

We should change this to only run on pull requests. Running stuff on `main` is unnecessary as our branch protection rules prevent us from merging if it doesn't pass anyway.

I guess it's good to run at least the _"Build & Test"_ one on `main`.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

### PR DESCRIPTION
**Pull request by Andarist**
Closes #3272

## 🎯 Changes

I think that this should suffice your needs but if anything is missing - let me know, I can tweak this further.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -1,6 +1,10 @@
 name: Build & Test
 
-on: [push]
+on:
+  push:
+    branches: main
+  pull_request:
+    branches: '**'
```
Comment (QuiiBz) [inline .github/workflows/main.yml:7]: We can remove this line if we want to run this workflow on every PRs
```suggestion
```

Comment (Andarist) [reply .github/workflows/main.yml:7]: That's what I did at first but a value-less field/property looked odd visually to me so I've added the value~ here. This should also execute for all branches

I don't mind adjusting it to the proposed version if that's what you prefer

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -1,6 +1,10 @@
...
 
 jobs:
   test:
```
Comment (Andarist) [inline .github/workflows/main.yml:10]: Actually, I've now noticed that this workflow depends on some secrets and those are not available to pull requests created by outside collaborators.

I think that Turbo should be able to fall back to just executing everything without using its remote cache (that's acceptable to me) and Codecov doesn't require the token for public repos... but you still provide it:
https://github.com/trpc/trpc/blob/22f268d1a16daa9154db2a456323366471fe30d7/.github/workflows/main.yml#L57

Can that explicit token just be removed?

**Comment by KATT** [comment]
Moving this to #3275

**Pull request closed without merging**